src/platform/backends/libvirt/libvirt_virtual_machine.cpp

```diff
@@ -345,41 +346,50 @@ void mp::LibVirtVirtualMachine::start()
     monitor->on_resume();
 }
 
-void mp::LibVirtVirtualMachine::shutdown()
+void mp::LibVirtVirtualMachine::shutdown(const bool force)
 {
-    std::unique_lock<decltype(state_mutex)> lock{state_mutex};
-    auto domain = domain_by_name_for(vm_name, open_libvirt_connection(libvirt_wrapper).get(), libvirt_wrapper);
+    std::unique_lock<std::mutex> lock{state_mutex};
+    auto domain = checked_vm_domain();
+
     state = refresh_instance_state_for_domain(domain.get(), state, libvirt_wrapper);
-    if (state == State::suspended)
+
+    try
     {
-        mpl::log(mpl::Level::info, vm_name, fmt::format("Ignoring shutdown issued while suspended"));
+        check_state_for_shutdown(force);
     }
-    else
+    catch (const VMStateIdempotentException& e)
     {
-        drop_ssh_session();
+        mpl::log(mpl::Level::info, vm_name, e.what());
+        return;
+    }
 
-        if (state == State::running || state == State::delayed_shutdown || state == State::unknown)
-        {
-            if (!domain || libvirt_wrapper->virDomainShutdown(domain.get()) == -1)
-            {
-                auto warning_string{
-                    fmt::format("Cannot shutdown '{}': {}", vm_name, libvirt_wrapper->virGetLastErrorMessage())};
-                mpl::log(mpl::Level::warning, vm_name, warning_string);
-                throw std::runtime_error(warning_string);
-            }
-
-            state = State::off;
-            update_state();
-        }
-        else if (state == State::starting)
+    if (force)
```

Comment:
```suggestion
    if (force) // TODO delete suspension state if it exists
```

Reply:
About the libvirt code, do we still want to bother to add the TODO given the fact that we will remove libvirt code soon?

Reply:
It is not a "fact" yet ;)